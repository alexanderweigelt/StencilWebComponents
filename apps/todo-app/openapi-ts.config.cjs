const apiSpecPath = require.resolve('backend/todo-api.yaml');

/** @type {import('@hey-api/openapi-ts').UserConfig} */
module.exports = {
  client: '@hey-api/client-fetch',
  input: apiSpecPath,
  output: 'src/generated/client',
};
