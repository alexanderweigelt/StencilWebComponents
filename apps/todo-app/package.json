{
  "name": "todo-app",
  "private": true,
  "version": "0.0.1",
  "description": "Stencil App Starter",
  "scripts": {
    "build": "stencil build",
    "start": "stencil build --dev --watch --serve",
    "test": "stencil test --spec --e2e",
    "test.watch": "stencil test --spec --e2e --watchAll",
    "generate": "stencil generate"
  },
  "devDependencies": {
    "@stencil/core": "^4.7.0",
    "@types/jest": "^29.5.6",
    "@types/node": "^16.18.11",
    "jest": "^29.7.0",
    "jest-cli": "^29.7.0",
    "puppeteer": "^21.9.0"
  },
  "dependencies": {
    "ui": "*"
  }
}
