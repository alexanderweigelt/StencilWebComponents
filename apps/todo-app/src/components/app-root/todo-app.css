header {
  background: #5851ff;
  padding: 12px;
  display: flex;
  align-items: center;
  box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26);
}

main {
  height: 100vh;
  padding: 12px;
  background-color: #f9f9f9;
}
