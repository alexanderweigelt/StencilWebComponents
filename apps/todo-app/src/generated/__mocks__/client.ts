export const getTodos = jest.fn();
export const postTodos = jest.fn();
