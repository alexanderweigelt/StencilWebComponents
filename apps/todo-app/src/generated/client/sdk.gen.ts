// This file is auto-generated by @hey-api/openapi-ts

import { createClient, createConfig, type OptionsLegacyParser } from '@hey-api/client-fetch';
import type { GetTodosError, GetTodosResponse, PostTodosData, PostTodosError, PostTodosResponse, GetTodosByIdData, GetTodosByIdError, GetTodosByIdResponse, PutTodosByIdData, PutTodosByIdError, PutTodosByIdResponse, DeleteTodosByIdData, DeleteTodosByIdError, DeleteTodosByIdResponse } from './types.gen';

export const client = createClient(createConfig());

/**
 * Hole eine Liste aller ToDos
 * Gibt eine Liste von ToDo-Elementen zurück, sortiert nach Datum (aufsteigend).
 */
export const getTodos = <ThrowOnError extends boolean = false>(options?: OptionsLegacyParser<unknown, ThrowOnError>) => {
    return (options?.client ?? client).get<GetTodosResponse, GetTodosError, ThrowOnError>({
        ...options,
        url: '/todos'
    });
};

/**
 * Erstelle ein neues ToDo
 * Fügt ein neues ToDo-Element zur Liste hinzu.
 */
export const postTodos = <ThrowOnError extends boolean = false>(options: OptionsLegacyParser<PostTodosData, ThrowOnError>) => {
    return (options?.client ?? client).post<PostTodosResponse, PostTodosError, ThrowOnError>({
        ...options,
        url: '/todos'
    });
};

/**
 * Hole ein spezifisches ToDo
 * Gibt ein einzelnes ToDo-Element anhand seiner ID zurück.
 */
export const getTodosById = <ThrowOnError extends boolean = false>(options: OptionsLegacyParser<GetTodosByIdData, ThrowOnError>) => {
    return (options?.client ?? client).get<GetTodosByIdResponse, GetTodosByIdError, ThrowOnError>({
        ...options,
        url: '/todos/{id}'
    });
};

/**
 * Aktualisiere ein ToDo
 * Aktualisiert ein bestehendes ToDo-Element anhand seiner ID.
 */
export const putTodosById = <ThrowOnError extends boolean = false>(options: OptionsLegacyParser<PutTodosByIdData, ThrowOnError>) => {
    return (options?.client ?? client).put<PutTodosByIdResponse, PutTodosByIdError, ThrowOnError>({
        ...options,
        url: '/todos/{id}'
    });
};

/**
 * Lösche ein ToDo
 * Entfernt ein ToDo-Element anhand seiner ID.
 */
export const deleteTodosById = <ThrowOnError extends boolean = false>(options: OptionsLegacyParser<DeleteTodosByIdData, ThrowOnError>) => {
    return (options?.client ?? client).delete<DeleteTodosByIdResponse, DeleteTodosByIdError, ThrowOnError>({
        ...options,
        url: '/todos/{id}'
    });
};