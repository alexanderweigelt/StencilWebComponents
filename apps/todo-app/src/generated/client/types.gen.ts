// This file is auto-generated by @hey-api/openapi-ts

export type NewToDo = {
    /**
     * Der Titel des ToDo-Elements.
     */
    text: string;
    /**
     * Markiert das ToDo-Element als erledigt.
     */
    completed: boolean;
    /**
     * Das Fälligkeitsdatum des ToDo-Elements.
     */
    dueDate: string;
};

export type ToDo = {
    /**
     * Die eindeutige ID des ToDo-Elements.
     */
    id: number;
    /**
     * Der Titel des ToDo-Elements.
     */
    text: string;
    /**
     * Markiert das ToDo-Element als erledigt.
     */
    completed: boolean;
    /**
     * Das Fälligkeitsdatum des ToDo-Elements.
     */
    dueDate: string;
};

export type GetTodosResponse = (Array<ToDo>);

export type GetTodosError = unknown;

export type PostTodosData = {
    body: NewToDo;
};

export type PostTodosResponse = (ToDo);

export type PostTodosError = unknown;

export type GetTodosByIdData = {
    path: {
        /**
         * Die ID des ToDo-Elements.
         */
        id: number;
    };
};

export type GetTodosByIdResponse = (ToDo);

export type GetTodosByIdError = (unknown);

export type PutTodosByIdData = {
    body: NewToDo;
    path: {
        /**
         * Die ID des ToDo-Elements.
         */
        id: number;
    };
};

export type PutTodosByIdResponse = (ToDo);

export type PutTodosByIdError = (unknown);

export type DeleteTodosByIdData = {
    path: {
        /**
         * Die ID des ToDo-Elements.
         */
        id: number;
    };
};

export type DeleteTodosByIdResponse = (void);

export type DeleteTodosByIdError = (unknown);