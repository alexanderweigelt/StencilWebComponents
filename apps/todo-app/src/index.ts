import 'ui/dist/todo-ui/todo-ui.esm.js'
export { Components, JSX } from './components';
