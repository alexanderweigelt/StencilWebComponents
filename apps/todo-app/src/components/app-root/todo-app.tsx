import { Component, Fragment, h, State, JSX, Host } from '@stencil/core';
import { getTodos, postTodos, ToDo } from '../../generated/client';
import { TodoList } from '../todo-list/todo-list';
import { AddTodo } from '../add-todo/add-todo';

@Component({
  tag: 'todo-app',
  styleUrl: 'todo-app.css',
  shadow: true,
})
export class TodoApp {
  @State() errorMessage: string | null = null;
  @State() loading: boolean = false;
  @State() todos: ToDo[] = [];
  @State() newTodoText: string = '';

  componentWillLoad() {
    this.fetchTodos();
  }

  /**
   * Saves a new ToDo
   */
  addTodo = async () => {
    if (this.newTodoText.trim() === '') return;

    try {
      const response = await postTodos({
        baseUrl: 'http://localhost:3000',
        body: { dueDate: Date.now().toLocaleString('de-DE'), text: this.newTodoText.trim(), completed: false },
      });
      this.todos = [...this.todos, response.data];
      this.newTodoText = '';
    } catch (error) {
      this.errorMessage = error instanceof Error ? error.message : 'An error occurred';
    } finally {
      this.loading = false;
    }
  };

  /**
   * Toggles a ToDo between done or open
   * @todo Add API fetch to save the todo state in a further step
   * @param id
   */
  toggleTodo = (id: number) => {
    // should fetch the ToDo API to set item as completed
    this.todos = this.todos.map(todo =>
      todo.id === id ? { ...todo, completed: !todo.completed } : todo,
    );
  };

  /**
   * Fetch ToDo API and returns a list of todos
   */
  fetchTodos = async () => {
    this.loading = true;
    this.errorMessage = null;

    try {
      const response = await getTodos({ baseUrl: 'http://localhost:3000' });
      if (Array.isArray(response.data)) {
        this.todos = response.data;
      } else {
        throw new Error('Unexpected API response format');
      }
    } catch (error) {
      this.errorMessage = error instanceof Error ? error.message : 'An error occurred';
    } finally {
      this.loading = false;
    }
  };

  handleInputChange = (event: Event) => {
    const input = event.target as HTMLInputElement;
    this.newTodoText = input.value;
  };

  private renderToDoApp = (): JSX.Element => {
    return (
      <Fragment>
        <TodoList todos={this.todos} toggleTodo={this.toggleTodo} />
        <AddTodo addTodo={this.addTodo} newTodoText={this.newTodoText} handleInputChange={this.handleInputChange} />
      </Fragment>
    );
  };

  render() {
    return (
      <Host>
        <header>
          <ui-headline>ToDo App</ui-headline>
        </header>

        <main>
          {this.loading && <p>ToDos list are loaded...</p>}
          {!this.loading && this.errorMessage && <p class="error">{this.errorMessage}</p>}
          {!this.loading && !this.errorMessage && this.renderToDoApp()}
        </main>
      </Host>
    );
  }
}
