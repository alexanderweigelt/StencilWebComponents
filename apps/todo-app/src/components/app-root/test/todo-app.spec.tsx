jest.mock('../../../generated/client')

import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { TodoApp } from '../todo-app';
import { ToDo, getTodos, postTodos } from '../../../generated/client';

describe('TodoApp Web-Component', () => {
  let page: SpecPage
  let instance: TodoApp;
  let mockToDos: ToDo[];

  beforeEach(async () => {
    page = await newSpecPage({
      components: [TodoApp],
      html: `<todo-app></todo-app>`,
    })
    await page.waitForChanges();
    instance = page.rootInstance
    mockToDos = [
      { id: 1, text: 'Learn Stencil', dueDate: '20.12.2024, 17:40:45', completed: false },
      { id: 2, text: 'Write Tests', dueDate: '24.12.2024, 18:45:48', completed: true },
    ];
  })

  it('renders the app correctly', async () => {
    (getTodos as jest.Mock).mockReturnValueOnce({ data: mockToDos });

    await page.rootInstance.componentWillLoad();
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('adds a new todo correctly', async () => {
    const mockNewTodo = { id: 3, text: 'New ToDo', completed: false, dueDate: '28.12.2024, 12:30:33' };
    (postTodos as jest.Mock).mockResolvedValueOnce({ data: mockNewTodo });

    instance.newTodoText = 'New ToDo';
    await instance.addTodo();
    await page.waitForChanges();

    expect(instance.todos).toContainEqual(mockNewTodo);
    expect(instance.newTodoText).toBe('');
  });

  it('toggles a todo status correctly', () => {
    instance.todos = mockToDos;

    instance.toggleTodo(1);
    expect(instance.todos[0].completed).toBe(true);
    expect(instance.todos[1].completed).toBe(true);
  });

  it('renders loading state correctly', async () => {
    instance.errorMessage = null;
    instance.loading = true;
    await page.waitForChanges();

    const loadingText = page.root.shadowRoot.querySelector('.loading');
    expect(loadingText.textContent).toBe('ToDos list are loaded...');
  });

  it('handles fetchTodos success state', async () => {
    (getTodos as jest.Mock).mockReturnValueOnce({ data: mockToDos });

    await instance.fetchTodos();
    await page.waitForChanges();

    expect(instance.todos).toEqual(mockToDos);
    expect(instance.loading).toBe(false);
    expect(page.root.shadowRoot.querySelector('.loading')).toBeNull();
  });

  it('handles fetchTodos error state', async () => {
    (getTodos as jest.Mock).mockReturnValueOnce(new Error('API Error'));

    await instance.fetchTodos();
    await page.waitForChanges();

    expect(instance.todos).toEqual([]);
    expect(instance.errorMessage).toBe('Unexpected API response format');
    expect(page.root.shadowRoot.querySelector('.loading')).toBeNull();
    const errorText = page.root.shadowRoot.querySelector('.error');
    expect(errorText.textContent).toBe('Unexpected API response format');
  });

  it('shows loading animation while fetching todos', async () => {
    (getTodos as jest.Mock).mockImplementation(() => {
      return new Promise(resolve =>
        setTimeout(() => resolve({ data: mockToDos }), 50)
      );
    });

    instance.fetchTodos();
    await page.waitForChanges();

    expect(instance.loading).toBe(true);
    const loadingText = page.root.shadowRoot.querySelector('.loading');
    expect(loadingText).not.toBeNull();
    expect(loadingText.textContent).toBe('ToDos list are loaded...');

    // Waiting for the completion of `fetchTodos`
    await new Promise(resolve => setTimeout(resolve, 60));

    expect(instance.loading).toBe(false);
    expect(instance.errorMessage).toBeNull();
  });

  it('removes loading animation after error', async () => {
    (getTodos as jest.Mock).mockRejectedValueOnce(new Error('API Error'));

    await instance.fetchTodos();
    await page.waitForChanges();

    expect(instance.loading).toBe(false);
    expect(instance.errorMessage).toBe('API Error');
    const loadingText = page.root.shadowRoot.querySelector('.loading');
    expect(loadingText).toBeNull();
  });
});
