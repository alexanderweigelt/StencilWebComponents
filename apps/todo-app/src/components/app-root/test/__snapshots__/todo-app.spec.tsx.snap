// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp Web-Component renders the app correctly 1`] = `
<todo-app>
  <template shadowrootmode="open">
    <header>
      <ui-headline>
        ToDo App
      </ui-headline>
    </header>
    <main class="todo-app">
      <ul class="todo-list">
        <li class="todo-item">
          <ui-button>
            open
          </ui-button>
          <ui-text>
            Learn Stencil
          </ui-text>
        </li>
        <li class="completed todo-item">
          <ui-button>
            complete
          </ui-button>
          <ui-text>
            Write Tests
          </ui-text>
        </li>
      </ul>
      <div class="add-todo">
        <form>
          <input id="new-todo" placeholder="Add a todo" value="">
          <button>
            Add Todo
          </button>
        </form>
      </div>
    </main>
  </template>
</todo-app>
`;
